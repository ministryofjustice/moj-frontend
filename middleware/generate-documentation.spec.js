const { generateMarkdown } = require('./generate-documentation')

describe('generateMarkdown', () => {
  beforeAll(() => {
    jest.useFakeTimers('modern')
    jest.setSystemTime(new Date('2023-01-01T00:00:00Z'))
  })

  afterAll(() => {
    jest.useRealTimers()
  })

  it('should generate markdown with all fields populated', () => {
    const mockData = {
      '/component-details': {
        componentName: 'Test Component',
        briefDescription: 'This is a brief description.',
        componentOverview: 'This is an overview.',
        componentProblemSolved: 'This is the problem solved.',
        howIsTheComponentUsed: 'This is how it is used.'
      },
      '/prototype-url': {
        prototypeUrl: 'www.test.com'
      },
      '/component-code-details': {
        componentCode: '<p>hello!</p>',
        componentCodeLanguage: 'html',
        howIsTheComponentUsed: 'Somehow'
      },
      '/additional-information': {
        additionalInformation: 'Additional information here.'
      },
      '/your-details': {
        fullName: 'Test User',
        emailAddress: 'test@test.com'
      },
      '/add-external-audit': {
        externalOrganisation: 'External Org',
        'auditDate-day': '1',
        'auditDate-month': '1',
        'auditDate-year': '2022',
        issuesDiscovered: 'Some issues discovered.',
        accessibilityReport: 'external-report.pdf'
      },
      '/add-internal-audit': {
        internalOrganisation: 'Internal Org',
        'auditDate-day': '2',
        'auditDate-month': '2',
        'auditDate-year': '2022',
        issuesDiscovered: 'Some internal issues discovered.',
        accessibilityReport: 'internal-report.pdf'
      },
      '/add-assistive-tech': {
        'testingDate-day': '3',
        'testingDate-month': '3',
        'testingDate-year': '2022',
        issuesDiscovered: 'Assistive tech issues discovered.',
        accessibilityReport: 'assistive-tech-report.pdf'
      }
    }
    const mockFiles = {
      '/component-image': {
        path: 'assets/images/component.jpg'
      },
      '/add-external-audit': {
        path: 'external-report.pdf'
      },
      '/add-internal-audit': {
        path: 'internal-report.pdf'
      },
      '/add-assistive-tech': {
        path: 'assistive-tech-report.pdf'
      }
    }

    const result = generateMarkdown(mockData, mockFiles)
    expect(result).toMatchSnapshot()
  })

  it('should handle missing details gracefully', () => {
    const mockData = {
      '/component-details': {},
      '/component-code-details': {},
      '/additional-information': {},
      '/your-details': {},
      '/add-external-audit': {},
      '/add-internal-audit': {},
      '/add-assistive-tech': {}
    }
    const mockFiles = {}

    const result = generateMarkdown(mockData, mockFiles)
    expect(result).toMatchSnapshot()
  })

  it('should handle missing image data gracefully', () => {
    const mockData = {
      '/component-details': {
        componentName: 'Another Component',
        briefDescription: 'Another brief description.',
        componentOverview: 'Another overview.',
        componentProblemSolved: 'Another problem solved.',
        howIsTheComponentUsed: 'Another usage.'
      },
      '/component-code-details': {
        componentCode: '<p>hello!</p>',
        componentCodeLanguage: 'html',
        howIsTheComponentUsed: 'Somehow'
      },
      '/additional-information': {
        additionalInformation: 'Additional information here.'
      },
      '/your-details': {
        fullName: 'Another User',
        emailAddress: 'another@test.com'
      },
      '/add-external-audit': {
        externalOrganisation: 'External Org',
        'auditDate-day': '1',
        'auditDate-month': '1',
        'auditDate-year': '2022',
        issuesDiscovered: 'Some issues discovered.',
        accessibilityReport: 'external-report.pdf'
      },
      '/add-internal-audit': {
        internalOrganisation: 'Internal Org',
        'auditDate-day': '2',
        'auditDate-month': '2',
        'auditDate-year': '2022',
        issuesDiscovered: 'Some internal issues discovered.',
        accessibilityReport: 'internal-report.pdf'
      },
      '/add-assistive-tech': {
        'testingDate-day': '3',
        'testingDate-month': '3',
        'testingDate-year': '2022',
        issuesDiscovered: 'Assistive tech issues discovered.',
        accessibilityReport: 'assistive-tech-report.pdf'
      }
    }
    const mockFiles = {
      '/add-external-audit': {
        path: 'external-report.pdf'
      },
      '/add-internal-audit': {
        path: 'internal-report.pdf'
      },
      '/add-assistive-tech': {
        path: 'assistive-tech-report.pdf'
      }
    }
    const result = generateMarkdown(mockData, mockFiles)
    expect(result).toMatchSnapshot()
  })

  it('should sanitize the filename correctly', () => {
    const mockData = {
      '/component-details': {
        componentName: 'Component! With@ Special#Chars',
        briefDescription: 'Brief description here.',
        componentOverview: 'Overview here.',
        componentProblemSolved: 'Problem solved here.',
        howIsTheComponentUsed: 'Usage here.'
      },
      '/component-code-details': {
        componentCode: '<p>hello!</p>',
        componentCodeLanguage: 'html',
        howIsTheComponentUsed: 'Somehow'
      },
      '/additional-information': {
        additionalInformation: 'Additional information here.'
      },
      '/your-details': {
        fullName: 'Special User',
        emailAddress: 'special@test.com'
      },
      '/add-external-audit': {
        externalOrganisation: 'External Org',
        'auditDate-day': '1',
        'auditDate-month': '1',
        'auditDate-year': '2022',
        issuesDiscovered: 'Some issues discovered.',
        accessibilityReport: 'external-report.pdf'
      },
      '/add-internal-audit': {
        internalOrganisation: 'Internal Org',
        'auditDate-day': '2',
        'auditDate-month': '2',
        'auditDate-year': '2022',
        issuesDiscovered: 'Some internal issues discovered.',
        accessibilityReport: 'internal-report.pdf'
      },
      '/add-assistive-tech': {
        'testingDate-day': '3',
        'testingDate-month': '3',
        'testingDate-year': '2022',
        issuesDiscovered: 'Assistive tech issues discovered.',
        accessibilityReport: 'assistive-tech-report.pdf'
      }
    }

    const result = generateMarkdown(mockData, {})
    expect(result).toMatchSnapshot()
  })

  it('should return minimal markdown if no valid data is provided', () => {
    const mockData = {
      '/component-details': {},
      '/component-code-details': {},
      '/additional-information': {},
      '/your-details': {},
      '/add-external-audit': {},
      '/add-internal-audit': {},
      '/add-assistive-tech': {}
    }

    const result = generateMarkdown(mockData)
    expect(result).toMatchSnapshot()
  })
})
