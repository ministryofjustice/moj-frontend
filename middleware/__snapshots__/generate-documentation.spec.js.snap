// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateMarkdown should generate markdown with all fields populated 1`] = `
{
  "content": "---
layout: layouts/tabbed-component.njk
title: Test Component
type: component
status: Experimental
statusDate: January 2023
eleventyNavigation:
  key: Test Component
  parent: Components
  excerpt: "This is a brief description."
---

{% tabs "Contents" %}

{% tab "Overview" %}

## Overview

<div class="img-container">
  ![Test Component](/assets/images/component.jpg)
</div>

This is an overview.

## How the component is being used

This is how it is used.

{% endtab %}

{% tab "Code" %}

## Code

No code has been added for this component.

If you've used this component in your service and you have code, you can add it. Most users are looking for HTML, Nunjucks, Javascript and CSS or SASS.

{% endtab %}

{% tab "Accessibility" %}

## Accessibility

If you have had an accessibility audit or tested with users with access needs then you could contribute to this component.

### External audit

Some issues discovered.

[Download the accessibility report](/external-report.pdf)

### Internal review

Some internal issues discovered.

[Download the accessibility report](/internal-report.pdf)

### Assistive Technology testing

{% endtab %}

{% tab "Prototypes and design files" %}

## Prototypes and design files

### Prototypes

<a href="www.test.com" target="_blank" rel="noopener noreferrer">Prototype link (opens in a new tab)</a>

{% endtab %}

{% endtabs %}
",
  "filename": "docs/components/test-component.md",
}
`;

exports[`generateMarkdown should handle missing details gracefully 1`] = `
{
  "content": "---
layout: layouts/tabbed-component.njk
title: unknown-component
type: component
status: Experimental
statusDate: January 2023
eleventyNavigation:
  key: unknown-component
  parent: Components
  excerpt: ""
---

{% tabs "Contents" %}

{% tab "Overview" %}

## Overview

<div class="img-container">

</div>

## How the component is currently used

{% endtab %}

{% tab "Code" %}

## Help develop existing building blocks in GitHub

After a new building block is published in the design system, you, and other users, have the chance to continue enhancing it. This is done with users adding more information and resources to the component via GitHub.

To do this you should:

- go to the GitHub conversation
- add your comments, information and resources about the building block

## Code

No code has been added for this component. If you have examples of how you have used this component in your service then you could help the community. Most users are looking for HTML, Nunjucks, Javascript and CSS or SASS.

{% endtab %}

{% tab "Accessibility" %}

## Accessibility

No accessibility findings have been added for this component.

{% endtab %}

{% tab "Prototypes and design files" %}

## Prototypes and design files

No prototypes or design files have been provided for this component. If you have used this component in your service and you have a prototype you can share it here.


{% endtab %}

{% endtabs %}
",
  "filename": "docs/components/unknown-component.md",
}
`;

exports[`generateMarkdown should handle missing image data gracefully 1`] = `
{
  "content": "---
layout: layouts/tabbed-component.njk
title: Another Component
type: component
status: Experimental
statusDate: January 2023
eleventyNavigation:
  key: Another Component
  parent: Components
  excerpt: "Another brief description."
---

{% tabs "Contents" %}

{% tab "Overview" %}

## Overview

<div class="img-container">

</div>

Another overview.

## How the component is being used

Another usage.

{% endtab %}

{% tab "Code" %}

## Code

## Code

No code has been added for this component. If you have examples of how you have used this component in your service then you could help the community. Most users are looking for HTML, Nunjucks, Javascript and CSS or SASS.

{% endtab %}

{% tab "Accessibility" %}

## Accessibility

This
If you have accessibility findings or tested with users with access needs then you could contribute to this component.

### External audit (External Org) - January 2022
Some issues discovered.

[Download the accessibility report](/external-report.pdf)

### Internal audit (Internal Org) - February 2022

Some internal issues discovered.

[Download the accessibility report](/internal-report.pdf)

### Assistive Technology audit - March 2022

Assistive tech issues discovered.

[Download the accessibility report](/assistive-tech-report.pdf)

{% endtab %}

{% tab "Prototypes and design files" %}

## Prototypes and design files

No prototypes or design files have been added for this component. If you have used this component in your service and you have a prototype or design file you can share it here.

{% endtab %}

{% endtabs %}
",
  "filename": "docs/components/another-component.md",
}
`;

exports[`generateMarkdown should return minimal markdown if no valid data is provided 1`] = `
{
  "content": "---
layout: layouts/tabbed-component.njk
title: unknown-component
type: component
status: Experimental
statusDate: January 2023
eleventyNavigation:
  key: unknown-component
  parent: Components
  excerpt: ""
---

{% tabs "Contents" %}

{% tab "Overview" %}

## Overview

<div class="img-container">

</div>

## How the component is currently being used

{% endtab %}

{% tab "Code" %}

## Code

No code has been added for this component. If you have examples of how you've used this component in your service then you could help the community by adding some. Most users are looking for HTML, Nunjucks, JavaScript, CSS or SASS.

{% endtab %}

{% tab "Accessibility" %}

## Accessibility

No accessibility findings have been added for this component. If you have examples of how you've used this component in your service then you could help the community by adding some. Most users are looking for HTML, Nunjucks, JavaScript, CSS or SASS.

{% endtab %}

{% tab "Prototype and design files" %}

## Prototypes and design files

No prototypes or design files have been added for this component. If you've used this component in your service and you have a prototype you can share it in the Github discussion.

{% endtab %}

{% endtabs %}
",
  "filename": "docs/components/unknown-component.md",
}
`;

exports[`generateMarkdown should sanitize the filename correctly 1`] = `
{
  "content": "---
layout: layouts/tabbed-component.njk
title: Component! With@ Special#Chars
type: component
status: Experimental
statusDate: January 2023
eleventyNavigation:
  key: Component! With@ Special#Chars
  parent: Components
  excerpt: "Brief description here."
---

{% tabs "Contents" %}

{% tab "Overview" %}

## Overview

<div class="img-container">

</div>

Overview here.

## How the component is being used

Usage here.

{% endtab %}

{% tab "Code" %}

## Code

No code has been added for this component. If you have examples of how you've used this component in your service then you could help the community by adding some. Most users are looking for HTML, Nunjucks, JavaScript, CSS or SASS.

{% endtab %}

{% tab "Accessibility" %}

## Accessibility

If you have had an accessibility audit or tested with users with access needs then you could contribute to this component.

### External audit

Some issues discovered.

### Internal review

Some internal issues discovered.

### Assistive Technology testing

Assistive tech issues discovered.

{% endtab %}

{% tab "Prototype and design files" %}

## Prototype and design files

No prototypes or design files have been provided for this component. If you have used this component in your service and you have a prototype you can share it here.

{% endtab %}

{% endtabs %}
",
  "filename": "docs/components/component-with-special-chars.md",
}
`;
